### PR TITLE
Add unit tests for Regorus rule file

### PR DESCRIPTION
**Pull request by mgunnala**
## Description

Add unit tests for Regorus rule file (rego file). 
Rename main class in regorus.py to "Regorus". 
Rename rule file to "agent_policy.rego" and remove unused rules.

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/ga/policy/policy_engine.py [hunk 1]
```diff
@@ -130,7 +129,7 @@ def evaluate_query(self, input_to_check, query):
             "extensions": {
                 "<extension_name_1>": {
                     "signingInfo": {
-                        "extensionSigned": <true, false>
+                        "extensionSigned": <True, False>
```
Comment (yangjie-msft) [inline azurelinuxagent/ga/policy/policy_engine.py:132]: Nit: in Json, Boolean value is represented in lower case.

Comment (mgunnala) [reply azurelinuxagent/ga/policy/policy_engine.py:132]: Expected format is a python dict. Booleans are capitalized in Python but converted to lowercase with the json.dump() call in eval_query().

File: tests/ga/test_regorus.py [hunk 2]
```diff
@@ -58,42 +72,294 @@ def tearDownClass(cls):
         cls.patcher.stop()
         AgentTestCase.tearDownClass()
 
-    def test_should_evaluate_query_with_valid_params(self):
+    def test_download_allowed_should_be_true_if_extension_in_allowlist(self):
+        """
+        If extension is in the allowlist, downloadAllowed=True. Global allowlist rule shouldn't matter.
```
Comment (yangjie-msft) [inline tests/ga/test_regorus.py:77]: NIT: maybe you can add comment on what rule you target to test

File: tests/ga/test_regorus.py [hunk 3]
```diff
@@ -58,42 +72,294 @@ def tearDownClass(cls):
...
+    def test_signing_validated_should_depend_on_signature_if_individual_signing_rule_true(self):
+        """
+        If individual signing rule is enforced, signingValidated=true only if extension is signed.
+        Global signing rule shouldn't matter.
+        """
+        # Should be same result for global signing rule true/false, so we test both cases.
+        global_signing_rule_cases = [True, False]
+        for global_signing_rule in global_signing_rule_cases:
+            policy = {
+                "azureGuestAgentPolicy": {
```
Comment (maddieford) [inline tests/ga/test_regorus.py:154]: This is where I'm having confusion about the 'extensionSigned' property. 
In the policy, what does 'extensionSigned': True mean? I assume it means that the agent should enforce signature validation. 

In the input, what does signingRules.extensionSigned == True for an extension mean? I assume it means that the extension itself is signed. 

If this is the case then these two properties have different meanings and it would be more readable if they were named differently

If i'm completely misunderstanding this lmk

File: tests/ga/test_regorus.py
Comment (maddieford) [inline tests/ga/test_regorus.py:1]: Should we add tests for the case when user-provided policy is empty or not provided, so we fall back on default policy?

File: tests/ga/test_policy_engine.py [hunk 4]
```diff
@@ -28,8 +27,22 @@ class TestPolicyEngine(AgentTestCase):
...
-    input_json = None  # Input is stored in a file, and extracted into this variable during class setup.
+    default_rule_path = os.path.join(data_dir, 'policy', "agent_policy.rego")
+    input_json = {
+        "extensions": {
+            "Microsoft.Azure.ActiveDirectory.AADSSHLoginForLinux": {
+                "signingInfo": {
+                    "extensionSigned": False
+                }
+            },
+            "test2": {
```
Comment (maddieford) [inline tests/ga/test_policy_engine.py:38]: Could we rename this to TestExtension2, same for test3 below

File: tests/data/policy/agent_policy.rego [hunk 5]
```diff
@@ -21,62 +21,50 @@ import rego.v1
...
 	global_rules.allowListOnly
 }
 
 default default_signing_info := {"signingInfo": {}}
 
-# Download rule 1: if the extension is in the list and download rule satisfied: download allowed
+# Download rule 1: if extension is in the allowlist, download is allowed.
 extensions_to_download[name] := extension if {
 	some name, input_extension in input.extensions
 	data.azureGuestExtensionsPolicy[name]
```
[the remark this answers is not in the document]
Comment (narrieta) [reply tests/data/policy/agent_policy.rego:56]: +1; also pointed out int he spec review

**Review by narrieta — COMMENTED** [review]
A few initial comments; review still in progress

File: tests/data/policy/agent_policy.rego [hunk 6]
```diff
@@ -21,62 +21,50 @@ import rego.v1
 
 policy_version := "0.1.0"
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:22]: Yes, we will increment this each time we update agent_policy.rego.

File: tests/data/policy/agent_policy.rego [hunk 7]
```diff
@@ -21,62 +21,50 @@ import rego.v1
...
-		"extensionSigned": false,
-		"signingDetails": {},
-	},
-	"updateAllowed": true,
-	"uninstallAllowed": true,
+		"extensionSigned": true
+	}
 }
 
+# Rules from policy file (data.azureGuestAgentPolicy) override the default rules if present.
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:39]: "data" variable holds the input data from the user - what we've been calling the "policy file" (agent_policy.json). data.azureGuestAgentPolicy refers to the section "azureGuestAgentPolicy" in that file.

File: tests/data/policy/agent_policy.rego [hunk 8]
```diff
@@ -21,62 +21,50 @@ import rego.v1
...
 
+# If allowlist rule is true, download is not allowed by default.
 any_extension_allowed := false if {
 	global_rules.allowListOnly
 }
 
 default default_signing_info := {"signingInfo": {}}
 
-# Download rule 1: if the extension is in the list and download rule satisfied: download allowed
+# Download rule 1: if extension is in the allowlist, download is allowed.
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:53]: Yes, this rule doesn't depend on allowListOnly. If the extension is in the allowlist, it's allowed regardless of the global allowlist rule.

File: tests/data/policy/agent_policy.rego [hunk 9]
```diff
@@ -21,62 +21,50 @@ import rego.v1
 
 policy_version := "0.1.0"
 
+# Defines default rules to be used if not specified in policy file.
 default default_global_rules := {
-	"allowListOnly": false,
+	"allowListOnly": true,
```
Comment (narrieta) [inline tests/data/policy/agent_policy.rego:26]: "allowListOnly" is confusing, since there isn't really an "allow list". There is a list of extensions, but that list does not act only as an allow list, it can be used to set conditions on the signature, irrespective of "allowListOnly"

File: tests/data/policy/agent_policy.rego [hunk 10]
```diff
@@ -21,62 +21,50 @@ import rego.v1
...
-		"signingDetails": {},
-	},
-	"updateAllowed": true,
-	"uninstallAllowed": true,
+		"extensionSigned": true
+	}
 }
 
+# Rules from policy file (data.azureGuestAgentPolicy) override the default rules if present.
 global_rules := object.union(default_global_rules, data.azureGuestAgentPolicy) if {
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:40]: Rego doesn't allow that kind of if/else logic, which is why it had to be implemented in this way. Each rego rule can evaluate to only one possible value.

File: tests/data/policy/agent_policy.rego [hunk 11]
```diff
@@ -21,62 +21,50 @@ import rego.v1
...
 	data.azureGuestAgentPolicy
 }
 
 default any_extension_allowed := true
 
+# If allowlist rule is true, download is not allowed by default.
 any_extension_allowed := false if {
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:47]: See comment above

File: tests/data/policy/agent_policy.rego [hunk 12]
```diff
@@ -100,7 +88,7 @@ extensions_validated[name] := extension if {
 	extension := object.union(output, {"signingValidated": false})
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:88]: We can change this to "signingPolicyValidated". It's trying to convey that the extension signature is validated according to the signing policy rules.

File: tests/data/policy/agent_policy.rego (hunk 6 above)
[the remark this answers is not in the document]
Comment (maddieford) [reply tests/data/policy/agent_policy.rego:22]: Update a rule? or update the file at all

[the remark this answers is not in the document]
Comment (maddieford) [reply tests/data/policy/agent_policy.rego:22]: Could you please add a comment

File: tests/data/policy/agent_policy.rego (hunk 12 above)
[the remark this answers is not in the document]
Comment (maddieford) [reply tests/data/policy/agent_policy.rego:88]: Does that mean the signature of the input extension is validated before evaluating whether it needs to be validated? Can’t we skip validation if not required by the rules?

File: tests/data/policy/agent_policy.rego (hunk 9 above)
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:26]: We could change this to "ListedExtensionsOnly"

File: tests/data/policy/agent_policy.rego (hunk 5 above)
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:56]: Eventually, if we wanted to add other non-extension policies to the single data file, it would go under the "azureGuestAgentPolicy" section (for example, JIT policy). So for now it's extensions only, but it will contain all policies globally affecting agent behavior. 

We could change it to "azureGuestAgentGlobalPolicy" and "azureGuestExtensionsIndividualPolicy", if that's clearer.

File: tests/data/policy/agent_policy.rego [hunk 13]
```diff
@@ -21,62 +21,50 @@ import rego.v1
...
-		"extensionSigned": false,
-		"signingDetails": {},
-	},
-	"updateAllowed": true,
-	"uninstallAllowed": true,
+		"extensionSigned": true
+	}
 }
 
 default global_rules := {
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:32]: We need to define a default in the case that the user provides an invalid or empty policy. Default rules cannot contain references or variables.

File: tests/data/policy/agent_policy.rego (hunk 5 above)
[the remark this answers is not in the document]
Comment (maddieford) [reply tests/data/policy/agent_policy.rego:56]: That naming makes more sense to me, but we should still update the properties in the global policy section to indicate they are related to extensions. 
'allowListOnly' isn't clear that this property is related to extensions for example

[the remark this answers is not in the document]
Comment (narrieta) [reply tests/data/policy/agent_policy.rego:56]: How about going from

```
{
   "azureGuestAgentPolicy": {
      "policyVersion": "0.1.0",
      "signingRules": {
         "extensionSigned": true
      },
      "allowListOnly": true
   },
   "azureGuestExtensionsPolicy": {
      "<extension name1>": {
         "runtimeRules": {
            "filesMustBeSigned": true,
            "fileRootCert": "<dirs>",
            "allowedCommands": ["<cmd1>, "<cmd2>" ]
         }
      }
   }
}

```
to something similar to

```
{
   "policyVersion": "0.1.0",
   "extensionPolicies": {
      "allowAllExtensions": false,
      "allowUnsignedExtensions": false
   },
   "<extension name1>": {
      "allowUnsigned": true,
      "fileRootCert": "<dirs>",
      "allowedCommands": [ "<cmd1>, "<cmd2>" ]
   }
}
```

File: tests/data/policy/agent_policy.rego (hunk 13 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply tests/data/policy/agent_policy.rego:32]: OK, please add a comment explaining that. It should help clarify why there is a "default default_global_rules" and a "default global_rules".  Also add a comment saying those 2 need to match.

File: tests/data/policy/agent_policy.rego (hunk 7 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply tests/data/policy/agent_policy.rego:39]: could you add that as a comment? ("data" variable holds the input data from the user - what we've been calling the "policy file" (agent_policy.json))

File: tests/data/policy/agent_policy.rego (hunk 8 above)
[the remark this answers is not in the document]
Comment (yangjie-msft) [reply tests/data/policy/agent_policy.rego:53]: the rego doesn't support if else, rule 2 is else condition of rule 1

File: tests/ga/test_regorus.py (hunk 3 above)
Comment (yangjie-msft) [reply tests/ga/test_regorus.py:154]: there are three parts of the policy: rego, data, and input. Rego defines how to compare data and input. In the policy data we define rules as extensionSigned which means extension has to be signed or not. this is in the tag of "signing rules". In input agent provides the fact about the extension in the tag of "signingInfo" via extensionSigned flag. In rego we compare whether the user provided signingInfo matches the data defined signingRule.

Comment (maddieford) [reply tests/ga/test_regorus.py:154]: Does the property signing rules need to have the same name as the data in signing info? I think the rule should be named something like 'allowUnsignedExtensions' as Norberto suggested and the data should be named 'extensionSigned'

Comment (maddieford) [reply tests/ga/test_regorus.py:154]: Having the same name for the rule and data, when they convey different things is hard to follow in the code

File: tests/data/policy/agent_policy.rego (hunk 12 above)
[the remark this answers is not in the document]
Comment (yangjie-msft) [reply tests/data/policy/agent_policy.rego:88]: now this tag here really means whether the signing policy has been enabled in the rule and whether the extension has met this policy (signed). According to our discussion, if it is signed, and policy didn't request that, we still validate the signature, but will not abort installing and running the extension. Will just log the error. Otherwise, exception will be thrown.

[the remark this answers is not in the document]
Comment (maddieford) [reply tests/data/policy/agent_policy.rego:88]: I think we should validate all signatures and collect telemetry, but once the feature is stable we should probably only validate when required by policy.

If this tag is indicating that the extension meets the requirement that it is signed (but not necessarily that the signature is validated) then I think we should change the name to indicate that

[the remark this answers is not in the document]
Comment (yangjie-msft) [reply tests/data/policy/agent_policy.rego:88]: yes, Manu and I suggest this tag is changed to "signingPolicyValidated"

[the remark this answers is not in the document]
Comment (maddieford) [reply tests/data/policy/agent_policy.rego:88]: Would 'signatureValidationRequired' make sense?

File: tests/data/policy/agent_policy.rego (hunk 5 above)
[the remark this answers is not in the document]
Comment (yangjie-msft) [reply tests/data/policy/agent_policy.rego:56]: I put signing related rules under the tag of "signingRules" because we could add additional signing validation rules in the future, for example certain certificate validation policy: (what time to use: current time or signing time, whether we validate certificate usage, etc. All these options were valid options when validating signature and certificate, but until we received requests from customers), we don't need to change the overall structure of Rego, rather changing the one validation function we have at the end of this rego file. Similarly, I put "fileRootCert", "allowedcommands" under "runtime_rules" because each extension may have its own settings which the agent rego rule doesn't need to understand. I was even thinking customer could encrypt these settings and put encryped string under the runtime settings, like how environment variables being passed to extension. For the detailed word choices of the tag names like allowAllExtensions and allowUnsignedExtensions , it may make the rules look more complicated. I need to experiment.  but it does look better. if they don't increase the complexity, I would like to follow the suggestion.

File: tests/data/policy/agent_policy.rego (hunk 12 above)
[the remark this answers is not in the document]
Comment (yangjie-msft) [reply tests/data/policy/agent_policy.rego:88]: it could be rule doesn't request signing, but the extension has been signed. or the rule doesn't request signing, and the extension is not signed, in both cases signingPolicyValidated will be true. if changed to signatureValidationRequired, it is doesn't cover the latter case.

File: tests/data/policy/agent_policy.rego (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:44]: Added this as comment

File: tests/data/policy/agent_policy.rego (hunk 13 above)
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:37]: Added these comments.

File: tests/data/policy/agent_policy.rego (hunk 10 above)
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:46]: Added a comment to explain this

File: tests/data/policy/agent_policy.rego [hunk 14]
```diff
@@ -100,7 +88,7 @@ extensions_validated[name] := extension if {
 	extension := object.union(output, {"signingValidated": false})
 }
 
-# Validate rule 3: if individual signing rule doesn't exist, signing rule validated according to global signing rule
+# Validate rule 3: if individual signing rule doesn't exist, signing rule validated according to global signing rule.
 extensions_validated[name] := extension if {
 	some name, input_extension in input.extensions
 	not data.azureGuestExtensionsPolicy[name]
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego:100]: Fixed

File: tests/data/policy/agent_policy.rego [hunk 15]
```diff
@@ -88,7 +76,7 @@ extensions_validated[name] := extension if {
 	extension := object.union(output, {"signingValidated": true})
 }
 
-# Validate rule 2: if indivual signing rule exists, signing rule not validated according to the rules
+# Validate rule 2: if indivual signing rule exists, signing rule is not validated according to the rules.
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/data/policy/agent_policy.rego]: updated

File: tests/ga/test_regorus.py [hunk 16]
```diff
@@ -18,18 +18,37 @@
...
-    input_json = None  # Input is stored in a file, and extracted into this variable during class setup.
+    default_rule_path = os.path.join(data_dir, 'policy', "agent_policy.rego")
+    default_input = {
+        "extensions": {
+            "Microsoft.Azure.ActiveDirectory.AADSSHLoginForLinux": {
+                "signingInfo": {
+                    "extensionSigned": False
+                }
+            },
+            "test2": {
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/ga/test_regorus.py]: updated here and below

File: tests/ga/test_regorus.py [hunk 17]
```diff
@@ -58,42 +72,294 @@ def tearDownClass(cls):
...
+                        "extensionSigned": False
+                    },
+                    "allowListOnly": allowlist_only
+                },
+                "azureGuestExtensionsPolicy": {
+                    ALLOWED_EXT: {}
+                }
+            }
+
+            input_dict = {
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/ga/test_regorus.py]: changed

File: tests/ga/test_regorus.py [hunk 18]
```diff
@@ -58,42 +72,294 @@ def tearDownClass(cls):
...
+                "extensions": {
+                    ALLOWED_EXT: {}  # Extension in allowlist
+                }
+            }
+
+            with tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=True) as policy_file:
+                json.dump(policy, policy_file, indent=4)
+                policy_file.flush()
+                engine = Regorus(policy_file.name, self.default_rule_path)
+                output = engine.eval_query(input_dict, "data.agent_extension_policy.extensions_to_download")
```
[the remark this answers is not in the document]
Comment (mgunnala) [reply tests/ga/test_regorus.py]: I've updated the return value of eval_query(). I've added TODOs for the expected output dict, I will update those after we finalize the rule file names offline.

File: tests/ga/test_regorus.py
Comment (mgunnala) [reply tests/ga/test_regorus.py:1]: "test_download_allowed_should_be_false_if_policy_invalid()" tests this--if the expected sections are missing in the file, Regorus essentially treats the policy as empty--but I added an explicit test case for empty policy as well.

File: tests/ga/test_policy_engine.py (hunk 4 above)
Comment (mgunnala) [reply tests/ga/test_policy_engine.py:38]: Done

File: tests/ga/test_regorus.py (hunk 2 above)
Comment (mgunnala) [reply tests/ga/test_regorus.py:78]: Done

File: tests/data/policy/agent_policy.rego (hunk 6 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply tests/data/policy/agent_policy.rego:23]: who checks that version? or is it for documentation purposes only?

**Comment by narrieta** [comment]
Evaluation of the policies will follow a different approach
